document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java

```diff
@@ -283,17 +284,54 @@ public boolean updateSubDoc(Key key, String subDocPath, Document subDocument) {
       UpdateResult writeResult =
           collection.updateOne(selectionCriteriaForKey(key), setObject, new UpdateOptions());
       if (LOGGER.isDebugEnabled()) {
-        LOGGER.debug("Write result: " + writeResult.toString());
+        LOGGER.debug("Write result: " + writeResult);
       }
       // TODO:look into the writeResult to ensure it was successful. Was not easy to find this from
       // docs.
       return true;
     } catch (IOException e) {
-      LOGGER.error("Exception inserting document. key: {} content:{}", key, subDocument);
+      LOGGER.error("Exception updating document. key: {} content:{}", key, subDocument);
       return false;
     }
   }
 
+  @Override
+  public int bulkUpdateSubDoc(Map<Key, Map<String, Document>> documents) {
+    List<UpdateManyModel<BasicDBObject>> bulkWriteUpdate = new ArrayList<>();
+    for (Key key : documents.keySet()) {
+      Map<String, Document> subDocuments = documents.get(key);
+      List<BasicDBObject> updateOperations = new ArrayList<>();
+      for (String subDocPath : subDocuments.keySet()) {
+        Document subDocument = subDocuments.get(subDocPath);
+        String jsonString = subDocument.toJson();
+        try {
+          JsonNode jsonNode = MAPPER.readTree(jsonString);
+          // escape "." and "$" in field names since Mongo DB does not like them
+          JsonNode sanitizedJsonNode = recursiveClone(jsonNode, this::encodeKey);
+          BasicDBObject dbObject =
+              new BasicDBObject(
+                  subDocPath, BasicDBObject.parse(MAPPER.writeValueAsString(sanitizedJsonNode)));
+          dbObject.append(LAST_UPDATED_TIME, System.currentTimeMillis());
+          BasicDBObject setObject = new BasicDBObject("$set", dbObject);
+          updateOperations.add(setObject);
+        } catch (Exception e) {
+          LOGGER.error("Exception updating document. key: {} content:{}", key, subDocument);
+          return -1;
+        }
+      }
+      bulkWriteUpdate.add(
+          new UpdateManyModel(selectionCriteriaForKey(key), updateOperations, new UpdateOptions()));
+    }
+    if (bulkWriteUpdate.isEmpty()) {
+      return 0;
+    }
+    BulkWriteResult writeResult = collection.bulkWrite(bulkWriteUpdate);
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug("Write result: " + writeResult);
+    }
+    return writeResult.getModifiedCount() + writeResult.getInsertedCount();
```

Comment:
I don't think it's inserting anything, right? 

Should it just be `writeResult.getModifiedCount()`?

Reply:
ok